sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/translation/SqlOperators.java

```diff
@@ -81,6 +72,12 @@
           x -> createTypeFactory().createSqlType(SqlTypeName.VARCHAR),
           new UdafImpl<>(new StringAgg.StringAggString()));
 
+  public static final SqlOperator ARR_AGG_ARR_FN =
+          createUdafOperator(
+                  "array_agg",
+                  x -> createTypeFactory().createSqlType(SqlTypeName.ARRAY),
```

Comment:
Here you simply created an array type without specifying the element type. I guess that's why the error message says the "inferred type is ARRAY NOT NULL" (just array type, your can ignore the "NOT NULL" suffix), where it should be "BIGINT NOT NULL ARRAY NOT NULL" (array of bigint).

To create an array type with element type specified, you may want to use `createTypeFactory().createArrayType()`, like here: https://github.com/robinyqiu/beam/blob/cbe87445d4259b6b485bc010231dda1895022d83/sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/ZetaSqlCalciteTranslationUtils.java#L170

Reply:
But `ARRAY_AGG` should support all element types, not just `INT64`. https://github.com/google/zetasql/blob/master/docs/aggregate_functions.md#array_agg

Reply:
@ibzib exactly. I try to implement generic array like not sure if it is the correct way.

 private static RelDataType relDataType =
          new JavaTypeFactoryImpl(RelDataTypeSystem.DEFAULT)
                  .builder()
                  .add("col_tinyint", SqlTypeName.TINYINT)
                  .add("col_smallint", SqlTypeName.SMALLINT)
                  .add("col_integer", SqlTypeName.INTEGER)
                  .add("col_bigint", SqlTypeName.BIGINT)
                  .add("col_float", SqlTypeName.FLOAT)
                  .add("col_double", SqlTypeName.DOUBLE)
                  .add("col_decimal", SqlTypeName.DECIMAL)
                  .add("col_string_varchar", SqlTypeName.VARCHAR)
                  .add("col_time", SqlTypeName.TIME)
                  .add("col_date", SqlTypeName.DATE)
                  .add("col_timestamp_with_local_time_zone", SqlTypeName.TIMESTAMP_WITH_LOCAL_TIME_ZONE)
                  .add("col_timestamp", SqlTypeName.TIMESTAMP)
                  .add("col_boolean", SqlTypeName.BOOLEAN)
                  .build();

  public static final SqlOperator ARR_AGG_ARR_FN =
          createUdafOperator(
                  "array_agg",
                  x -> createTypeFactory().createArrayType(**relDataType**, -1),
                  new UdafImpl<>(new ArrayAgg.ArrayAggArray()));

Reply:
Yeah, ARRAY_AGG should be generic. I said the type should be "BIGINT NOT NULL ARRAY NOT NULL" because that's from the error message.

Sonam, I believe there is some way that you can get the element type from the input `x`. You don't need to define your new mapping.

Reply:
x -> createTypeFactory().createArrayType(x.getOperandType(0), -1) worked.

Thanks @robinyqiu  @ibzib